// pilpat-mcp-security/dist/pii-redaction.js
/**
 * PII (Personally Identifiable Information) Redaction Module
 *
 * Detects and redacts sensitive personal information from text:
 * - Email addresses
 * - Phone numbers (US and Polish formats)
 * - Credit card numbers
 * - Social Security Numbers (SSN)
 * - PESEL (Polish National ID Numbers)
 * - Polish ID Cards (Dowód Osobisty)
 * - Polish Passports
 * - Bank account numbers
 * - Driver's license numbers
 * - Passport numbers
 * - IP addresses
 * - URLs with credentials
 *
 * Note: Public business data (NIP, REGON) and postal codes are NOT redacted
 * as they are publicly available and often required as input for legitimate queries.
 */
const DEFAULT_OPTIONS = {
    redactEmails: false, // Changed to false - emails often needed for legitimate business purposes
    redactPhones: true,
    redactCreditCards: true,
    redactSSN: true,
    redactPESEL: true,
    redactPolishIdCard: true,
    redactPolishPassport: true,
    redactPolishPhones: true,
    redactBankAccounts: true,
    redactDriverLicense: true,
    redactPassport: true,
    redactIPAddresses: true,
    redactURLCredentials: true,
    placeholder: '[REDACTED]',
};
/**
 * PII Detection Patterns
 *
 * Based on industry-standard regex patterns and OWASP guidelines
 * Includes Polish-specific patterns for the Polish market
 */
const PII_PATTERNS = {
    /**
     * Email addresses (RFC 5322 simplified)
     * Matches: user@example.com, first.last+tag@domain.co.uk
     */
    email: /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/g,
    /**
     * Phone numbers (multiple formats)
     * Matches: (123) 456-7890, +1-234-567-8900, 1234567890
     */
    phone: /(\+?1[-.\s]?)?\(?([0-9]{3})\)?[-.\s]?([0-9]{3})[-.\s]?([0-9]{4})\b/g,
    /**
     * Credit card numbers (Luhn algorithm would validate further)
     * Matches: 4532-1234-5678-9010, 4532123456789010
     */
    creditCard: /\b(?:\d{4}[-\s]?){3}\d{4}\b|\b\d{16}\b/g,
    /**
     * Social Security Numbers (SSN)
     * Matches: 123-45-6789, 123456789 (but not obvious test numbers)
     */
    ssn: /\b(?!000|666|9\d{2})\d{3}-?\d{2}-?\d{4}\b/g,
    /**
     * PESEL (Polish National ID Number)
     * Format: YYMMDDNNNCS (11 digits)
     * - YY: Year (last 2 digits)
     * - MM: Month (01-12 with century modifier: +20 for 2000-2099)
     * - DD: Day (01-31)
     * - NNN: Serial number
     * - C: Check digit
     * - S: Gender digit (odd=male, even=female)
     *
     * Examples: 44051401359 (1944-05-14, male), 02261012345 (2002-06-10, female)
     */
    pesel: /\b[0-9]{2}[0-3][0-9][0-3][0-9][0-9]{5}\b/g,
    /**
     * Polish ID Card (Dowód Osobisty)
     * Format: 3 letters + 6 digits
     *
     * Examples: ABC123456, CBA654321
     */
    polishIdCard: /\b[A-Z]{3}\d{6}\b/g,
    /**
     * Polish Passport Number
     * Format: 2 letters + 7 digits
     *
     * Examples: AB1234567, ZZ9876543
     */
    polishPassport: /\b[A-Z]{2}\d{7}\b/g,
    /**
     * Polish Phone Numbers
     * Format: +48 XXX XXX XXX (9 digits after country code)
     * - Mobile: starts with 4, 5, 6, 7, 8
     * - Landline: area code (2 digits) + 7 digits
     *
     * Examples: +48 123 456 789, +48-123-456-789, +48123456789, 0048 123 456 789
     */
    polishPhone: /(?:\+48|0048)[-\s]?\d{3}[-\s]?\d{3}[-\s]?\d{3}\b/g,
    /**
     * Bank account numbers (typically 10-12 digits)
     * Matches: account numbers in various formats
     */
    bankAccount: /\b\d{10,12}\b/g,
    /**
     * Driver's license numbers (varies by state, common pattern: 6-8 alphanumeric)
     * Matches: various state formats
     */
    driverLicense: /\b[A-Z]{1,2}\d{5,8}[A-Z]?\b/g,
    /**
     * Passport numbers (typically 6-9 alphanumeric)
     * Matches: international passport format
     */
    passport: /\b[A-Z]{1,2}\d{6,8}\b/g,
    /**
     * IPv4 addresses
     * Matches: 192.168.1.1 (excludes private ranges that might be internal)
     */
    ipv4: /\b(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\b/g,
    /**
     * IPv6 addresses
     * Matches: 2001:0db8:85a3:0000:0000:8a2e:0370:7334
     */
    ipv6: /(?:[0-9a-fA-F]{0,4}:){2,7}[0-9a-fA-F]{0,4}/g,
    /**
     * URLs with credentials (http://user:password@host)
     * Matches: https://admin:password123@example.com
     */
    urlCredentials: /(?:https?|ftp):\/\/[^\s:/@]+:[^\s/@]+@[^\s/]+/g,
};
/**
 * Redact Personally Identifiable Information (PII) from text
 *
 * @param text - The text containing potential PII
 * @param options - Redaction options
 * @returns Object with redacted text, detected PII types, and details
 *
 * @example
 * ```typescript
 * const result = redactPII(
 *   'Contact me at john@example.com or 555-123-4567',
 *   { redactEmails: true, redactPhones: true }
 * );
 * // result.redacted: 'Contact me at [REDACTED] or [REDACTED]'
 * // result.detectedPII: ['email', 'phone']
 * ```
 */
export function redactPII(text, options = {}) {
    if (typeof text !== 'string') {
        text = String(text);
    }
    const opts = { ...DEFAULT_OPTIONS, ...options };
    let redacted = text;
    const detectionDetails = {
        emails: [],
        phones: [],
        creditCards: [],
        ssns: [],
        peselNumbers: [],
        polishIdCards: [],
        polishPassports: [],
        polishPhones: [],
        bankAccounts: [],
        driverLicenses: [],
        passports: [],
        ipAddresses: [],
    };
    // Pattern matching order is CRITICAL: Most specific patterns FIRST, general patterns LAST
    // This prevents conflicts where general patterns (phone, bank account) match specific ones (PESEL, NIP)
    // Email addresses
    if (opts.redactEmails) {
        const matches = redacted.match(PII_PATTERNS.email) || [];
        const placeholder = opts.placeholder || '[REDACTED]';
        detectionDetails.emails = [...new Set(matches)];
        redacted = redacted.replace(PII_PATTERNS.email, placeholder);
    }
    // Credit card numbers (16 digits - very specific)
    if (opts.redactCreditCards) {
        const matches = redacted.match(PII_PATTERNS.creditCard) || [];
        // For credit cards, only keep last 4 digits in redaction info
        detectionDetails.creditCards = matches.map(cc => {
            const lastFour = cc.replace(/\D/g, '').slice(-4);
            return `****${lastFour}`;
        });
        redacted = redacted.replace(PII_PATTERNS.creditCard, opts.placeholder + ' (card)');
    }
    // PESEL (Polish National ID Numbers) - 11 digits, check BEFORE bank accounts
    if (opts.redactPESEL) {
        const matches = redacted.match(PII_PATTERNS.pesel) || [];
        const placeholder = opts.placeholder || '[REDACTED]';
        detectionDetails.peselNumbers = matches.map(() => placeholder);
        redacted = redacted.replace(PII_PATTERNS.pesel, placeholder + ' (PESEL)');
    }
    // Social Security Numbers (9 digits - specific format)
    if (opts.redactSSN) {
        const matches = redacted.match(PII_PATTERNS.ssn) || [];
        detectionDetails.ssns = matches.map(() => opts.placeholder);
        redacted = redacted.replace(PII_PATTERNS.ssn, opts.placeholder + ' (SSN)');
    }
    // Polish ID Cards (3 letters + 6 digits)
    if (opts.redactPolishIdCard) {
        const matches = redacted.match(PII_PATTERNS.polishIdCard) || [];
        const placeholder = opts.placeholder || '[REDACTED]';
        detectionDetails.polishIdCards = matches.map(() => placeholder);
        redacted = redacted.replace(PII_PATTERNS.polishIdCard, placeholder + ' (Polish ID)');
    }
    // Polish Passports (2 letters + 7 digits)
    if (opts.redactPolishPassport) {
        const matches = redacted.match(PII_PATTERNS.polishPassport) || [];
        const placeholder = opts.placeholder || '[REDACTED]';
        detectionDetails.polishPassports = matches.map(() => placeholder);
        redacted = redacted.replace(PII_PATTERNS.polishPassport, placeholder + ' (Polish passport)');
    }
    // Polish Phone Numbers (with +48 prefix)
    if (opts.redactPolishPhones) {
        const matches = redacted.match(PII_PATTERNS.polishPhone) || [];
        const placeholder = opts.placeholder || '[REDACTED]';
        detectionDetails.polishPhones = [...new Set(matches)];
        redacted = redacted.replace(PII_PATTERNS.polishPhone, placeholder + ' (PL phone)');
    }
    // US/International Phone numbers (10 digits - check AFTER Polish phones)
    if (opts.redactPhones) {
        const matches = redacted.match(PII_PATTERNS.phone) || [];
        const placeholder = opts.placeholder || '[REDACTED]';
        detectionDetails.phones = [...new Set(matches)];
        redacted = redacted.replace(PII_PATTERNS.phone, placeholder);
    }
    // Bank account numbers
    if (opts.redactBankAccounts) {
        const matches = redacted.match(PII_PATTERNS.bankAccount) || [];
        detectionDetails.bankAccounts = matches.map(() => opts.placeholder);
        redacted = redacted.replace(PII_PATTERNS.bankAccount, opts.placeholder + ' (account)');
    }
    // Driver's license numbers
    if (opts.redactDriverLicense) {
        const matches = redacted.match(PII_PATTERNS.driverLicense) || [];
        detectionDetails.driverLicenses = matches.map(() => opts.placeholder);
        redacted = redacted.replace(PII_PATTERNS.driverLicense, opts.placeholder + ' (DL)');
    }
    // Passport numbers
    if (opts.redactPassport) {
        const matches = redacted.match(PII_PATTERNS.passport) || [];
        detectionDetails.passports = matches.map(() => opts.placeholder);
        redacted = redacted.replace(PII_PATTERNS.passport, opts.placeholder + ' (passport)');
    }
    // IP addresses (both IPv4 and IPv6)
    if (opts.redactIPAddresses) {
        const ipv4Matches = redacted.match(PII_PATTERNS.ipv4) || [];
        const ipv6Matches = redacted.match(PII_PATTERNS.ipv6) || [];
        detectionDetails.ipAddresses = [...new Set([...ipv4Matches, ...ipv6Matches])];
        redacted = redacted.replace(PII_PATTERNS.ipv4, opts.placeholder + ' (IPv4)');
        redacted = redacted.replace(PII_PATTERNS.ipv6, opts.placeholder + ' (IPv6)');
    }
    // URL credentials
    if (opts.redactURLCredentials) {
        const matches = redacted.match(PII_PATTERNS.urlCredentials) || [];
        // Log URL detection (don't expose full URL in details)
        if (matches.length > 0) {
            detectionDetails.emails.push(...matches.map(() => opts.placeholder));
        }
        redacted = redacted.replace(PII_PATTERNS.urlCredentials, opts.placeholder + ' (URL credentials)');
    }
    // Determine detected PII types
    const detectedPII = [];
    if (detectionDetails.emails.length > 0)
        detectedPII.push('email');
    if (detectionDetails.phones.length > 0)
        detectedPII.push('phone');
    if (detectionDetails.creditCards.length > 0)
        detectedPII.push('credit_card');
    if (detectionDetails.ssns.length > 0)
        detectedPII.push('ssn');
    if (detectionDetails.peselNumbers.length > 0)
        detectedPII.push('pesel');
    if (detectionDetails.polishIdCards.length > 0)
        detectedPII.push('polish_id_card');
    if (detectionDetails.polishPassports.length > 0)
        detectedPII.push('polish_passport');
    if (detectionDetails.polishPhones.length > 0)
        detectedPII.push('polish_phone');
    if (detectionDetails.bankAccounts.length > 0)
        detectedPII.push('bank_account');
    if (detectionDetails.driverLicenses.length > 0)
        detectedPII.push('driver_license');
    if (detectionDetails.passports.length > 0)
        detectedPII.push('passport');
    if (detectionDetails.ipAddresses.length > 0)
        detectedPII.push('ip_address');
    return {
        redacted,
        detectedPII,
        detectionDetails,
    };
}
/**
 * Check if text contains any PII
 *
 * @param text - Text to check
 * @returns True if any PII patterns detected
 */
export function hasPII(text) {
    const result = redactPII(text);
    return result.detectedPII.length > 0;
}
/**
 * Get summary of PII found in text
 *
 * @param text - Text to analyze
 * @returns Summary of PII detection
 */
export function getPIISummary(text) {
    const result = redactPII(text);
    const count = Object.values(result.detectionDetails).reduce((sum, arr) => sum + arr.length, 0);
    return {
        hasPII: result.detectedPII.length > 0,
        piiTypes: result.detectedPII,
        count,
        summary: `Found ${count} PII items (${result.detectedPII.join(', ')})`,
    };
}
