// pilpat-mcp-security/dist/sanitization.js
/**
 * Output Sanitization Module
 *
 * Removes potentially dangerous content from text responses:
 * - HTML/XML tags and script content
 * - Control characters
 * - Dangerous Unicode
 * - Length enforcement
 */
const DEFAULT_OPTIONS = {
    removeHtml: true,
    maxLength: undefined,
    allowedTags: [],
    removeControlChars: true,
    normalizeWhitespace: true,
    removeEncodingPatterns: true,
};
/**
 * Common dangerous HTML tags that should always be removed
 */
const DANGEROUS_TAGS = [
    'script',
    'iframe',
    'object',
    'embed',
    'applet',
    'meta',
    'link',
    'style',
    'form',
    'input',
    'button',
    'onclick',
    'onerror',
    'onload',
    'onmouseover',
];
/**
 * Regex patterns for common encoding/obfuscation attempts
 */
const ENCODING_PATTERNS = [
    /&#x[0-9a-fA-F]+;/g, // Hex entities like &#x3C;
    /&#\d+;/g, // Decimal entities like &#60;
    /\\u[0-9a-fA-F]{4}/g, // Unicode escapes like \u003C
    /\\x[0-9a-fA-F]{2}/g, // Hex escapes like \x3C
];
/**
 * Sanitize output text by removing potentially dangerous content
 *
 * @param text - The text to sanitize
 * @param options - Sanitization options
 * @returns Sanitized text
 *
 * @example
 * ```typescript
 * const raw = '<script>alert("XSS")</script>Hello World';
 * const safe = sanitizeOutput(raw, { removeHtml: true });
 * // Result: "Hello World"
 * ```
 */
export function sanitizeOutput(text, options = {}) {
    if (typeof text !== 'string') {
        text = String(text);
    }
    const opts = { ...DEFAULT_OPTIONS, ...options };
    let sanitized = text;
    // 1. Remove encoding patterns first (they might hide dangerous content)
    if (opts.removeEncodingPatterns) {
        ENCODING_PATTERNS.forEach(pattern => {
            sanitized = sanitized.replace(pattern, '');
        });
    }
    // 2. Remove HTML/XML tags
    if (opts.removeHtml) {
        // Remove dangerous tags unconditionally
        DANGEROUS_TAGS.forEach(tag => {
            const regex = new RegExp(`<${tag}[^>]*>.*?</${tag}>`, 'gis');
            sanitized = sanitized.replace(regex, '');
        });
        // Remove all tags if no allowed tags specified
        if (!opts.allowedTags || opts.allowedTags.length === 0) {
            sanitized = sanitized.replace(/<[^>]+>/g, '');
        }
        else {
            // Remove only non-allowed tags
            const allowedTagsRegex = opts.allowedTags.join('|');
            const regex = new RegExp(`<(?!/?(?:${allowedTagsRegex})(?:\\s|>))[^>]+>`, 'gi');
            sanitized = sanitized.replace(regex, '');
        }
    }
    // 3. Remove control characters (except newline, carriage return, tab)
    if (opts.removeControlChars) {
        // eslint-disable-next-line no-control-regex
        sanitized = sanitized.replace(/[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]/g, '');
    }
    // 4. Normalize whitespace
    if (opts.normalizeWhitespace) {
        sanitized = sanitized
            .replace(/\r\n/g, '\n') // Normalize line endings
            .replace(/[ \t]+/g, ' ') // Collapse multiple spaces/tabs
            .replace(/\n{3,}/g, '\n\n'); // Collapse multiple newlines
    }
    // 5. Enforce maximum length
    if (opts.maxLength && sanitized.length > opts.maxLength) {
        sanitized = sanitized.substring(0, opts.maxLength - 3) + '...';
    }
    return sanitized;
}
/**
 * Detect if text contains common HTML/script injection patterns
 *
 * @param text - Text to check
 * @returns True if injection patterns detected
 */
export function hasInjectionPatterns(text) {
    const injectionPatterns = [
        /<script[^>]*>/i,
        /javascript:/i,
        /on\w+\s*=/i, // Event handlers like onclick=
        /<iframe/i,
        /<embed/i,
        /<object/i,
        /eval\(/i,
        /expression\(/i,
    ];
    return injectionPatterns.some(pattern => pattern.test(text));
}
/**
 * Get sanitization statistics (for monitoring)
 *
 * @param original - Original text
 * @param sanitized - Sanitized text
 * @returns Statistics about sanitization
 */
export function getSanitizationStats(original, sanitized) {
    return {
        originalLength: original.length,
        sanitizedLength: sanitized.length,
        removedLength: original.length - sanitized.length,
        removalPercentage: ((original.length - sanitized.length) / original.length) * 100,
        hadDangerousPatterns: hasInjectionPatterns(original),
    };
}
